File: src/validator.py
import os
import pandas as pd
import numpy as np
from utils import logger

REQUIRED_FIELDS = ['Date', 'Channel', 'CampaignType', 'CampaignName', 'Cost', 'Impressions', 'Clicks', 'Conversions', 'Revenue']

def validate_historical_data(df):
    """
    Validates dataset integrity and detects advertising campaign structural anomalies.
    Returns: (is_valid, errors_list, warnings_list, anomalies_list)
    """
    is_valid = True
    errors = []
    warnings = []
    anomalies = []
    
    if df is None or len(df) == 0:
        return False, ["No data found or CSV could not be parsed."], [], []
        
    # Header check
    missing_fields = [f for f in REQUIRED_FIELDS if f not in df.columns]
    if missing_fields:
        return False, [f"Missing required columns: {', '.join(missing_fields)}"], [], []
        
    # Row by row diagnostics
    df_clean = df.copy()
    
    # 1. Check Date column
    try:
        df_clean['Date'] = pd.to_datetime(df_clean['Date'])
    except Exception as e:
        errors.append(f"Invalid Date column values: {str(e)}")
        is_valid = False
        
    # Check negative cost/revenue
    negative_costs = df_clean[df_clean['Cost'] < 0]
    if not negative_costs.empty:
        errors.append(f"Cost cannot be negative (found {len(negative_costs)} negative rows)")
        is_valid = False
        
    negative_revenue = df_clean[df_clean['Revenue'] < 0]
    if not negative_revenue.empty:
        errors.append(f"Revenue cannot be negative (found {len(negative_revenue)} negative rows)")
        is_valid = False
        
    # Unique Channels & Naming variations check
    channels = df_clean['Channel'].unique()
    variations = ['google', 'meta', 'facebook', 'microsoft', 'bing']
    for v in variations:
        matches = [c for c in channels if v in str(c).lower()]
        if len(matches) > 1:
            warnings.append(f"Channel naming inconsistency: found duplicate variations for '{v}' ({', '.join(matches)})")
            
    # Calculate dataset timespan
    if is_valid:
        min_date = df_clean['Date'].min()
        max_date = df_clean['Date'].max()
        days_span = (max_date - min_date).days + 1
        if days_span < 60:
            warnings.append(f"Dataset covers only {days_span} days. Seasonality calculations and accuracy may be limited. Recommend at least 90+ days of history.")
            
    # Anomalies Detection
    for idx, row in df_clean.iterrows():
        line_num = idx + 2 # 1-based, +1 for header
        cost = float(row['Cost'])
        revenue = float(row['Revenue'])
        conversions = int(row['Conversions'])
        clicks = int(row['Clicks'])
        impressions = int(row['Impressions'])
        date_str = str(row['Date']).split(' ')[0]
        channel = row['Channel']
        campaign = row['CampaignName']
        
        # 1. Pixel/tracking failure: high spend, 0 conversions/revenue
        if cost > 300.0 and conversions == 0 and revenue == 0.0:
            anomalies.append({
                'date': date_str,
                'channel': channel,
                'campaign': campaign,
                'type': 'Tracking/Pixel Breakdown',
                'description': f"Campaign '{campaign}' spent ₹{cost:.2f} on {date_str} but reported 0 conversions and ₹0 revenue. Potential pixel outage.",
                'severity': 'High'
            })
            
        # 2. ROAS collapse / bid engine glitch
        if cost > 200.0 and revenue > 0.0 and (revenue / cost) < 0.15:
            roas = revenue / cost
            anomalies.append({
                'date': date_str,
                'channel': channel,
                'campaign': campaign,
                'type': 'ROAS Collapse / Bid Glitch',
                'description': f"Campaign '{campaign}' spent ₹{cost:.2f} on {date_str} but returned only ₹{revenue:.2f} revenue ({roas:.2f}x ROAS). Potential bid engine anomaly.",
                'severity': 'Medium'
            })
            
        # 3. CTR anomaly: click fraud or double tracking
        if impressions > 50 and clicks > 0:
            ctr = clicks / impressions
            if ctr > 0.45:
                anomalies.append({
                    'date': date_str,
                    'channel': channel,
                    'campaign': campaign,
                    'type': 'Abnormal CTR (Potential Click Fraud)',
                    'description': f"Campaign '{campaign}' reported {clicks} clicks out of {impressions} impressions ({(ctr * 100):.1f}% CTR) on {date_str}.",
                    'severity': 'Low'
                })
                
        # 4. Cost with zero impressions
        if cost > 50.0 and impressions == 0 and clicks > 0:
            anomalies.append({
                'date': date_str,
                'channel': channel,
                'campaign': campaign,
                'type': 'Incomplete Delivery Metrics',
                'description': f"Campaign '{campaign}' recorded {clicks} clicks and spent ₹{cost:.2f} on {date_str} but reported 0 impressions.",
                'severity': 'Low'
            })
            
    return is_valid, errors, warnings, anomalies

def run_diagnostics(file_path):
    """Utility runner to diagnose CSV file."""
    logger.info(f"Diagnosing data file: {file_path}")
    if not os.path.exists(file_path):
        logger.error(f"File not found: {file_path}")
        return False
        
    try:
        df = pd.read_csv(file_path)
        is_valid, errors, warnings, anomalies = validate_historical_data(df)
        
        logger.info(f"Integrity Check: {'PASSED' if is_valid else 'FAILED'}")
        for err in errors:
            logger.error(f"[Error] {err}")
        for warn in warnings:
            logger.warning(f"[Warning] {warn}")
            
        if anomalies:
            logger.warning(f"Detected {len(anomalies)} anomalies in data source:")
            for a in anomalies[:5]:  # print first 5
                logger.warning(f" - [{a['severity']}] {a['type']} on {a['date']} in {a['campaign']}: {a['description']}")
            if len(anomalies) > 5:
                logger.warning(f" ... and {len(anomalies)-5} more anomalies.")
                
        return is_valid
    except Exception as e:
        logger.error(f"Error reading file {file_path}: {str(e)}")
        return False

File: validator.js
/**
 * AdCast AI - CSV Parser, Data Validator, and Anomaly Detector
 * Handles file ingestion diagnostics and campaign structural validation client-side.
 */
(function() {
  const DataValidator = {
    // Required fields in the dataset
    requiredFields: ['Date', 'Channel', 'CampaignType', 'CampaignName', 'Cost', 'Impressions', 'Clicks', 'Conversions', 'Revenue'],

    /**
     * Parse CSV string into an array of objects
     * @param {string} csvText 
     * @returns {Array<Object>} parsed rows
     */
    parseCSV(csvText) {
      if (!csvText || !csvText.trim()) return [];
      
      const lines = csvText.split(/\r?\n/);
      if (lines.length < 2) return [];

      // Parse headers
      const headers = lines[0].split(',').map(h => h.trim().replace(/^["']|["']$/g, ''));
      const parsedData = [];

      for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        // Simple CSV splitter handling quoted values
        const matches = line.match(/(".*?"|[^",\s]+)(?=\s*,|\s*$)/g) || line.split(',');
        const row = {};
        
        headers.forEach((header, index) => {
          let val = matches[index] !== undefined ? matches[index].trim() : '';
          // Strip quotes
          val = val.replace(/^["']|["']$/g, '');
          row[header] = val;
        });

        // Convert numeric columns
        if (row.Cost !== undefined) row.Cost = parseFloat(row.Cost) || 0;
        if (row.Impressions !== undefined) row.Impressions = parseInt(row.Impressions, 10) || 0;
        if (row.Clicks !== undefined) row.Clicks = parseInt(row.Clicks, 10) || 0;
        if (row.Conversions !== undefined) row.Conversions = parseInt(row.Conversions, 10) || 0;
        if (row.Revenue !== undefined) row.Revenue = parseFloat(row.Revenue) || 0;

        parsedData.push(row);
      }

      return parsedData;
    },

    /**
     * Validates dataset integrity and tags structural warnings/anomalies
     * @param {Array<Object>} parsedData 
     * @returns {Object} diagnostic report
     */
    validate(parsedData) {
      const report = {
        isValid: true,
        errors: [],
        warnings: [],
        anomalies: [],
        stats: {
          totalRows: parsedData.length,
          totalSpend: 0,
          totalRevenue: 0,
          channels: new Set(),
          campaignTypes: new Set(),
          campaigns: new Set(),
          dateRange: { min: null, max: null }
        }
      };

      if (!parsedData || parsedData.length === 0) {
        report.isValid = false;
        report.errors.push("No data found or CSV could not be parsed.");
        return report;
      }

      // Check headers on first row
      const firstRowKeys = Object.keys(parsedData[0]);
      const missingFields = this.requiredFields.filter(f => !firstRowKeys.includes(f));
      
      if (missingFields.length > 0) {
        report.isValid = false;
        report.errors.push(`Missing required columns: ${missingFields.join(', ')}`);
        return report;
      }

      // Track dates to find min/max
      let minDate = null;
      let maxDate = null;

      parsedData.forEach((row, idx) => {
        const lineNum = idx + 2; // 1-based, +1 for header
        
        // Date validation
        const dateVal = row.Date;
        if (!dateVal) {
          report.errors.push(`Row ${lineNum}: Missing Date`);
          report.isValid = false;
        } else {
          const d = new Date(dateVal);
          if (isNaN(d.getTime())) {
            report.errors.push(`Row ${lineNum}: Invalid Date format "${dateVal}"`);
            report.isValid = false;
          } else {
            const time = d.getTime();
            if (!minDate || time < minDate.getTime()) minDate = d;
            if (!maxDate || time > maxDate.getTime()) maxDate = d;
          }
        }

        // Channel tracking & validation
        if (!row.Channel) {
          report.errors.push(`Row ${lineNum}: Missing Channel`);
          report.isValid = false;
        } else {
          report.stats.channels.add(row.Channel);
        }

        // Campaign structure tracking
        if (!row.CampaignName) {
          report.warnings.push(`Row ${lineNum}: CampaignName is empty`);
        } else {
          report.stats.campaigns.add(row.CampaignName);
        }

        if (!row.CampaignType) {
          report.warnings.push(`Row ${lineNum}: Campaign "${row.CampaignName}" has empty CampaignType`);
        } else {
          report.stats.campaignTypes.add(row.CampaignType);
        }

        // Value checking & anomalies detection
        if (row.Cost < 0) {
          report.errors.push(`Row ${lineNum}: Cost cannot be negative (${row.Cost})`);
          report.isValid = false;
        }
        if (row.Revenue < 0) {
          report.errors.push(`Row ${lineNum}: Revenue cannot be negative (${row.Revenue})`);
          report.isValid = false;
        }

        // Accumulate statistics
        report.stats.totalSpend += row.Cost;
        report.stats.totalRevenue += row.Revenue;

        // Anomaly Detection
        // 1. Pixel/tracking anomaly: zero conversions/revenue but high ad spend
        if (row.Cost > 300 && row.Conversions === 0 && row.Revenue === 0) {
          report.anomalies.push({
            date: row.Date,
            channel: row.Channel,
            campaign: row.CampaignName,
            type: 'Tracking/Pixel Breakdown',
            description: `Campaign "${row.CampaignName}" spent $${row.Cost.toFixed(2)} on ${row.Date} but reported 0 conversions and $0 revenue. Potential tracking pixel outage.`,
            severity: 'High'
          });
        }

        // 2. High CPC / Efficiency collapse: Cost is high, but revenue is extremely low (ROAS < 0.1x)
        if (row.Cost > 200 && row.Revenue > 0 && (row.Revenue / row.Cost) < 0.15) {
          const roas = row.Revenue / row.Cost;
          report.anomalies.push({
            date: row.Date,
            channel: row.Channel,
            campaign: row.CampaignName,
            type: 'ROAS Collapse / Bid Glitch',
            description: `Campaign "${row.CampaignName}" spent $${row.Cost.toFixed(2)} on ${row.Date} but returned only $${row.Revenue.toFixed(2)} revenue (${roas.toFixed(2)}x ROAS). Potential bid engine anomaly.`,
            severity: 'Medium'
          });
        }

        // 3. CTR anomaly: click-through rate over 40% (could indicate click fraud or double tracking)
        if (row.Impressions > 50 && row.Clicks > 0) {
          const ctr = row.Clicks / row.Impressions;
          if (ctr > 0.45) {
            report.anomalies.push({
              date: row.Date,
              channel: row.Channel,
              campaign: row.CampaignName,
              type: 'Abnormal CTR (Potential Click Fraud)',
              description: `Campaign "${row.CampaignName}" reported ${row.Clicks} clicks out of ${row.Impressions} impressions (${(ctr * 100).toFixed(1)}% CTR) on ${row.Date}.`,
              severity: 'Low'
            });
          }
        }

        // 4. Missing impressions with click spending
        if (row.Cost > 50 && row.Impressions === 0 && row.Clicks > 0) {
          report.anomalies.push({
            date: row.Date,
            channel: row.Channel,
            campaign: row.CampaignName,
            type: 'Incomplete Delivery Metrics',
            description: `Campaign "${row.CampaignName}" recorded ${row.Clicks} clicks and spent $${row.Cost} on ${row.Date} but reported 0 impressions. Check platform sync.`,
            severity: 'Low'
          });
        }
      });

      // Format date range
      if (minDate && maxDate) {
        report.stats.dateRange = {
          min: minDate.toISOString().split('T')[0],
          max: maxDate.toISOString().split('T')[0],
          days: Math.round((maxDate - minDate) / (1000 * 60 * 60 * 24)) + 1
        };
      }

      // Check naming consistency warnings
      const channelNames = Array.from(report.stats.channels);
      const variations = ['google', 'meta', 'facebook', 'microsoft', 'bing'];
      variations.forEach(v => {
        const matches = channelNames.filter(c => c.toLowerCase().includes(v));
        if (matches.length > 1) {
          report.warnings.push(`Channel naming inconsistency: found duplicate variations for "${v}" (${matches.join(', ')}). Please consolidate naming.`);
        }
      });

      // Check if dataset represents enough history (should ideally be at least 60-90 days for seasonal trends)
      if (report.stats.dateRange.days && report.stats.dateRange.days < 60) {
        report.warnings.push(`Dataset covers only ${report.stats.dateRange.days} days. Seasonality adjustments and forecasting accuracy may be limited. Recommend at least 90+ days of history.`);
      }

      return report;
    }
  };

  window.DataValidator = DataValidator;
})();
